### PR TITLE
feat: replaced flaky argon2 with better argon2 crate

### PR DESCRIPTION
**Pull request by NinekoTheCat**
fixes: https://github.com/girlbossceo/conduwuit/issues/22

### CODE REVIEW

File: src/service/globals/mod.rs [hunk 1]
```diff
@@ -188,7 +190,11 @@ impl Service {
             RoomVersionId::V5,
             RoomVersionId::V11,
         ];
-
+        let argon = Argon2::new(
+            argon2::Algorithm::Argon2id,
+            argon2::Version::default(),
+            argon2::Params::new(2432, 2, 1, None).expect("valid parameters"),
```
Comment (girlbossceo) [inline src/service/globals/mod.rs:196]: Can we add a comment what these config parameters are? I'm guessing `2432` is kilobytes, `2` is iterations, and `1` is amount of parallelism.

I'd like to use the following from https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#argon2id

> m=19456 (19 MiB), t=2, p=1 (Do not use with Argon2i)

File: src/api/client_server/session.rs [hunk 2]
```diff
@@ -74,9 +75,15 @@ pub async fn login_route(body: Ruma<login::v3::Request>) -> Result<login::v3::Re
                     "The user has been deactivated",
                 ));
             }
-
-            let hash_matches = argon2::verify_encoded(&hash, password.as_bytes()).unwrap_or(false);
-
+            let Ok(parsed_hash) = PasswordHash::new(&hash) else {
+                error!("error while hashing");
```
Comment (girlbossceo) [inline src/api/client_server/session.rs:79]: Can we log a bit more useful information to the admin? Something like

```rs
                error!("error while hashing user {}", user_id);
```

**Review by girlbossceo — CHANGES_REQUESTED** [review]
Just a couple small things, not as involved as I thought it'd be.

Also run `cargo fmt`

File: src/service/globals/mod.rs (hunk 1 above)
Comment (NinekoTheCat) [reply src/service/globals/mod.rs:196]: this is the one actually, I should add a comment yeah

Comment (girlbossceo) [reply src/service/globals/mod.rs:197]: Are you sure that's the one? 2432 kilobytes is ~2.5mb, that isn't 19MiB.

Comment (NinekoTheCat) [reply src/service/globals/mod.rs:197]: Kib != Kb

Comment (NinekoTheCat) [reply src/service/globals/mod.rs:197]: https://docs.rs/argon2/latest/argon2/struct.Params.html#arguments

Comment (girlbossceo) [reply src/service/globals/mod.rs:197]: It's still not 19MiB.

> m_cost: memory size in 1 KiB blocks. Between 1 and (2^32)-1.

image

Comment (NinekoTheCat) [reply src/service/globals/mod.rs]: sorry I fixed it, motherfucking wolfram got high and misinterpreted Mib as MebiBits

**Review by girlbossceo — APPROVED** [review]
Seems to work with no issues (user log outs, invalid passwords, etc). Don't see any obvious issues with code.

**Pull request merged**